plugin/rpcplugin/sandbox/sandbox_linux.go

```diff
@@ -0,0 +1,468 @@
+	if mountPoint.Type == "" {
+		flags |= syscall.MS_BIND
+	}
+	if mountPoint.ReadOnly {
+		flags |= syscall.MS_RDONLY
+	}
+
+	if err := syscall.Mount(mountPoint.Source, target, mountPoint.Type, flags, ""); err != nil {
+		return errors.Wrapf(err, "unable to mount "+mountPoint.Source)
+	}
+
+	if (flags & syscall.MS_BIND) != 0 {
+		// If this was a bind mount, our other flags actually got silently ignored during the above syscall:
+		//
+		//     If mountflags includes MS_BIND [...] The remaining bits in the mountflags argument are
+		//     also ignored, with the exception of MS_REC.
+		//
+		// Furthermore, remounting will fail if we attempt to unset a bit that was inherited from
+		// the mount's parent:
+		//
+		//     The mount(2) flags MS_RDONLY, MS_NOSUID, MS_NOEXEC, and the "atime" flags
+		//     (MS_NOATIME, MS_NODIRATIME, MS_RELATIME) settings become locked when propagated from
+		//     a more privileged to a less privileged mount namespace, and may not be changed in the
+		//     less privileged mount namespace.
+		//
+		// So we need to get the actual flags, add our new ones, then do a remount if needed.
+		var stats syscall.Statfs_t
+		if err := syscall.Statfs(target, &stats); err != nil {
+			return errors.Wrap(err, "unable to get mount flags for target: "+target)
+		}
+		const lockedFlagsMask = unix.MS_RDONLY | unix.MS_NOSUID | unix.MS_NOEXEC | unix.MS_NOATIME | unix.MS_NODIRATIME | unix.MS_RELATIME
+		lockedFlags := uintptr(stats.Flags & lockedFlagsMask)
+		if lockedFlags != ((flags | lockedFlags) & lockedFlagsMask) {
+			if err := syscall.Mount("", target, "", flags|lockedFlags|syscall.MS_REMOUNT, ""); err != nil {
+				return errors.Wrapf(err, "unable to remount "+mountPoint.Source)
+			}
+		}
+	}
+
+	return nil
+}
+
+func mountMountPoints(root string, mountPoints []*MountPoint) error {
+	for _, mountPoint := range mountPoints {
+		if err := mountMountPoint(root, mountPoint); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func pivotRoot(newRoot string) error {
+	if err := syscall.Mount(newRoot, newRoot, "", syscall.MS_BIND|syscall.MS_REC, ""); err != nil {
+		return errors.Wrapf(err, "unable to mount new root")
+	}
+
+	prevRoot := filepath.Join(newRoot, ".prev_root")
+
+	if err := os.MkdirAll(prevRoot, 0700); err != nil {
+		return errors.Wrapf(err, "unable to create directory for previous root")
+	}
+
+	if err := syscall.PivotRoot(newRoot, prevRoot); err != nil {
+		return errors.Wrapf(err, "syscall error")
+	}
+
+	if err := os.Chdir("/"); err != nil {
+		return errors.Wrapf(err, "unable to change directory")
+	}
+
+	prevRoot = "/.prev_root"
+
+	if err := syscall.Unmount(prevRoot, syscall.MNT_DETACH); err != nil {
+		return errors.Wrapf(err, "unable to unmount previous root")
+	}
+
+	if err := os.RemoveAll(prevRoot); err != nil {
+		return errors.Wrapf(err, "unable to remove previous root directory")
+	}
+
+	return nil
+}
+
+func dropInheritableCapabilities() error {
```

Comment:
The purpose of this is to drop any root capabilities in the case that MM is being run by root?

Reply:
Even if Mattermost isn't run by root, inside the sandbox, the plugin is run by the root user and has all capabilities (within the sandbox). So it can do a lot of things that we really don't need plugins to be able to do. For the most part, these capabilities wouldn't be able to influence the world outside of the sandbox. But some (such as CAP_NET_BIND_SERVICE since the network isn't namespaced and that's commonly granted to Mattermost) can, and they all represent increased attack surface.